### DISCUSSION TITLE
1) Constraining powers to integers, 2) Optimal number of cores to use with pySR

### DISCUSSION THREAD

**Discussion by festabc** [MilesCranmer/PySR#290] [General]
Hi Miles,

I hope you are doing well! 

I have two questions about pySR that came up during my research:
   1) Is there a way I can constrain the powers in an equation search to only integer numbers (or variables)?
   2) What is the default number of processors that pySR uses? If my computer has 8 cores (MacBookAir 10,1 ; AppleM1 chip), what is the optimal number of cores I should allow pySR to use?

Thanks so much in advance for your help!

Best,
Festa

**Comment by MilesCranmer** [comment]
Hi @festabc,

Great questions!

1. Yes, you can do this by defining a custom power law operator that converts its exponent to an integer:

```python
import sympy

PySRRegressor(
    binary_operators=[..., "pow_int(x, y) = pow(x, round(y))"]
    extra_sympy_operators={"pow_int": lambda x, y: x ** sympy.ceiling(y-0.5)},
    ....
)
```

2. It automatically should use all the cores (it sets `procs=cpu_count()`). So in your case, 8 should be good. If it doesn't occupy them all, you could try a higher number to saturate it.

Cheers!
Miles

**Reply by MilesCranmer** [reply]
Edit: I accidentally wrote `^` when it should have been `**` in the sympy version.

**Comment by festabc** [comment]
Thank you very much for your answer Miles! This is very helpful indeed!
